Handle macros with multiple kinds, and improve errors

joshtriplett opened the pull request:
(I recommend reviewing this commit-by-commit.)

Switch to a bitflags `MacroKinds` to support macros with more than one kind

Review everything that uses `MacroKind`, and switch anything that could refer to more than one kind to use `MacroKinds`.

Add a new `SyntaxExtensionKind::MacroRules` for `macro_rules!` macros, using the concrete `MacroRulesMacroExpander` type, and have it track which kinds it can handle. Eliminate the separate optional `attr_ext`, now that a `SyntaxExtension` can handle multiple macro kinds.

This also avoids the need to downcast when calling methods on `MacroRulesMacroExpander`, such as `get_unused_rule`.

Integrate macro kind checking into name resolution's `sub_namespace_match`, so that we only find a macro if it's the right type, and eliminate the special-case hack for attributes.

This allows detecting and report macro kind mismatches early, and more precisely, improving various error messages. In particular, this eliminates the case in `failed_to_match_macro` to check for a function-like invocation of a macro with no function-like rules.
    
Instead, macro kind mismatches now result in an unresolved macro, and we detect this case in `unresolved_macro_suggestions`, which now carefully distinguishes between a kind mismatch and other errors.
    
This also handles cases of forward-referenced attributes and cyclic attributes.
    
----

In this PR, I've minimally fixed up `rustdoc` so that it compiles and passes tests. This is just the minimal necessary fixes to handle the switch to `MacroKinds`, and it only works for macros that don't actually have multiple kinds. This will panic (with a `todo!`) if it encounters a macro with multiple kinds.
    
rustdoc needs further fixes to handle macros with multiple kinds, and to handle attributes and derive macros that aren't proc macros. I'd appreciate some help from a rustdoc expert on that.

----

r? @petrochenkov

joshtriplett committed:
Switch to a bitflags `MacroKinds` to support macros with more than one kind

Review everything that uses `MacroKind`, and switch anything that could
refer to more than one kind to use `MacroKinds`.

Add a new `SyntaxExtensionKind::MacroRules` for `macro_rules!` macros,
using the concrete `MacroRulesMacroExpander` type, and have it track
which kinds it can handle. Eliminate the separate optional `attr_ext`,
now that a `SyntaxExtension` can handle multiple macro kinds.

This also avoids the need to downcast when calling methods on
`MacroRulesMacroExpander`, such as `get_unused_rule`.

Integrate macro kind checking into name resolution's
`sub_namespace_match`, so that we only find a macro if it's the right
type, and eliminate the special-case hack for attributes.

joshtriplett committed:
Only suggest changing `#[derive(mymacro)]` to `#[mymacro]` for attribute macros

joshtriplett committed:
Detect and report macro kind mismatches early, and more precisely

This eliminates the case in `failed_to_match_macro` to check for a
function-like invocation of a macro with no function-like rules.

Instead, macro kind mismatches now result in an unresolved macro, and we
detect this case in `unresolved_macro_suggestions`, which now carefully
distinguishes between a kind mismatch and other errors.

This also handles cases of forward-referenced attributes and cyclic
attributes.

Expand test coverage to include all of these cases.

joshtriplett committed:
Expand documentation of `GlobDelegation`

I discovered this via research through the git log, and I want to leave
additional guidance for future macro spelunkers.

joshtriplett committed:
mbe: Rename macro parsing state names to use `Bang` instead of `Not`

The use of `Not` to describe the `!` in `macro_rules!` reads
confusingly, and also results in search collisions with the diagnostic
structure `MacroRulesNot` elsewhere in the compiler. Rename it to use
the more conventional `Bang` for `!`.

joshtriplett committed:
rustdoc: Minimal fixes to compile with `MacroKinds`

This makes the minimal fixes necessary for rustdoc to compile and pass
existing tests with the switch to `MacroKinds`. It only works for macros
that don't actually have multiple kinds, and will panic (with a `todo!`)
if it encounters a macro with multiple kinds.

rustdoc needs further fixes to handle macros with multiple kinds, and to
handle attributes and derive macros that aren't proc macros.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16847177340/job/47728360416) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47728360416) [(plain)](https://github.com/rust-lang/rust/commit/df2de34e7613688bc81236d0bc1d18d844363079/checks/47728360416/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] askama test:false 0.189
error[E0308]: mismatched types
   --> src/tools/clippy/clippy_lints/src/item_name_repetitions.rs:506:75
    |
506 |             let is_macro_rule = matches!(item.kind, ItemKind::Macro(_, _, MacroKind::Bang));
    |                                          ---------                        ^^^^^^^^^^^^^^^ expected `MacroKinds`, found `MacroKind`
    |                                          |
    |                                          this expression has type `rustc_hir::ItemKind<'_>`

error[E0308]: mismatched types
  --> src/tools/clippy/clippy_lints/src/redundant_pub_crate.rs:92:79
   |
92 |         let ignore = matches!(path.res.macro_ns, Some(Res::Def(DefKind::Macro(MacroKind::Bang), _)))
   |                               -----------------                               ^^^^^^^^^^^^^^^ expected `MacroKinds`, found `MacroKind`
   |                               |
   |                               this expression has type `std::option::Option<Res>`

For more information about this error, try `rustc --explain E0308`.
[RUSTC-TIMING] clippy_lints test:false 7.063
```

</details>

joshtriplett committed:
clippy: Update for switch to `MacroKinds`

This updates two clippy lints which had exceptions for `MacroKind::Bang`
macros to extend those exceptions to any macro, now that a macro_rules
macro can be any kind of macro.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16847498670/job/47729128020) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47729128020) [(plain)](https://github.com/rust-lang/rust/commit/1854363580af9de2088f83be6129d3adbb99cd2f/checks/47729128020/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

theemathas commented:
The cyclic macro from https://github.com/rust-lang/rust/pull/144579#issuecomment-3127590317 still gets a weird diagnostic. The `use` line is necessary. (`use`-ing a macro_rules, for reasons I don't quite understand, allows using the macro before it's defined.)

theemathas commented:
Is it intentional that I can define an attribute macro_rules and a function-like macro_rules, and use both independently?

This code compiles:
```rust
#![feature(macro_attr)]

macro_rules! foo {
    attr() { $($x:tt)* } => {};
}

macro_rules! foo {
    () => {};
}

#[foo]
struct Thing;

foo!();
```

petrochenkov commented on compiler/rustc_span/src/hygiene.rs:
```diff
@@ -1188,6 +1188,53 @@ impl MacroKind {
+/// A set of macro kinds, for macros that can have more than one kind
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Encodable, Decodable, Hash, Debug)]
+#[derive(HashStable_Generic)]
+pub struct MacroKinds(u8);
```
The structure isn't actually used in `rustc_span`.
Maybe move it to some higher crate in the dependency tree?

joshtriplett replied:
@petrochenkov It seems like it would be a little confusing for `MacroKind` and `MacroKinds` to be in two different crates.

That said, the four crates that use `MacroKinds` are `rustc_expand`, `rustc_hir`, `rustc_metadata`, and `rustc_resolve`.
- `rustc_expand` depends on `rustc_hir`
- `rustc_metadata` depends on `rustc_hir` and `rustc_expand`
- `rustc_resolve` depends on `rustc_hir` and `rustc_expand` and `rustc_metadata`

So, of those, the only one that might make sense is `rustc_hir`. I can move it there.

petrochenkov commented on compiler/rustc_lint/src/non_local_def.rs:
```diff
@@ -240,7 +240,7 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
                 )
             }
-            ItemKind::Macro(_, _macro, MacroKind::Bang)
+            ItemKind::Macro(_, _macro, _kind)
```
```suggestion
            ItemKind::Macro(_, _macro, _kinds)
```

petrochenkov commented on compiler/rustc_span/src/hygiene.rs:
```diff
@@ -1188,6 +1188,53 @@ impl MacroKind {
+            Self::BANG => "macro",
+            Self::ATTR => "attribute macro",
+            Self::DERIVE => "derive macro",
+            _ if self == (Self::ATTR | Self::BANG) => "attribute/function macro",
```
```suggestion
            const { Self::ATTR | Self::BANG } => "attribute/function macro",
```
should already work, IIRC.

joshtriplett replied:
@petrochenkov 

Not yet:

```text
error: const blocks cannot be used as patterns
    --> compiler/rustc_span/src/hygiene.rs:1223:19
     |
1223 |             const { Self::ATTR | Self::BANG } => "attribute/function macro",
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
```

(And even if that were allowed, bitflags' `BitOrd` implementation isn't `[const]`.)

fmease replied:
(inline const pats where removed: https://github.com/rust-lang/rust/pull/138492)

joshtriplett commented:
I've addressed all the feedback, and rebased to fix the conflict.

@rustbot ready

rust-log-analyzer commented:
The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16894616626/job/47861824113) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47861824113) [(plain)](https://github.com/rust-lang/rust/commit/a07625789240cc1dede85d0d01725db9b763b575/checks/47861824113/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16894830276/job/47862433182) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47862433182) [(plain)](https://github.com/rust-lang/rust/commit/861b36e173c9bc7fae2965423b94521c95933738/checks/47862433182/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

petrochenkov commented on compiler/rustc_hir/src/lib.rs:
```diff
@@ -39,6 +39,7 @@ pub mod weak_lang_items;
 #[cfg(test)]
 mod tests;
 
+pub use def::MacroKinds;
```
Any specific reasons for this?
In other files I mostly see `MacroKinds` imported from `rustc_hir::def`, and it would be better to have just one "canonical location" for the type.

joshtriplett replied:
@petrochenkov It's not required, it just seemed like a reasonable re-export given some of the other re-exports in `rustc_hir`, such as `pub use target::{MethodKind, Target};`.

I don't care strongly either way, though.

petrochenkov commented:
r=me after addressing [XXX ](https://github.com/rust-lang/rust/pull/145153#discussion_r2270401337) and squashing some history to avoid things like moving `MacroKinds`.
@rustbot author

bors commented:
:pushpin: Commit e1fc89af5bb657acb45097cae15873de78210065 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov e1fc89af5bb657acb45097cae15873de78210065 -->
<!-- homu: {"type":"Approved","sha":"e1fc89af5bb657acb45097cae15873de78210065","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-timer commented:
Finished benchmarking commit (c3f8df39f723e1bae9b21ee801f19e7fc46b1336): [comparison URL](https://perf.rust-lang.org/compare.html?start=f5b8a11fecf23cb669f6d1feae81aafc2930aeb9&end=c3f8df39f723e1bae9b21ee801f19e7fc46b1336&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f5b8a11fecf23cb669f6d1feae81aafc2930aeb9&end=c3f8df39f723e1bae9b21ee801f19e7fc46b1336&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 1.4%]   | 37    |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.6%]   | 20    |
| Improvements ✅ <br /> (primary)   | -0.8% | [-3.4%, -0.2%] | 6     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-3.4%, 1.4%]  | 43    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f5b8a11fecf23cb669f6d1feae81aafc2930aeb9&end=c3f8df39f723e1bae9b21ee801f19e7fc46b1336&stat=max-rss)
<details>
<summary>Results (secondary 4.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.6% | [4.6%, 4.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f5b8a11fecf23cb669f6d1feae81aafc2930aeb9&end=c3f8df39f723e1bae9b21ee801f19e7fc46b1336&stat=cycles:u)
<details>
<summary>Results (primary -0.8%, secondary -5.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8%  | [1.8%, 1.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.4% | [-3.4%, -3.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -5.0% | [-5.0%, -5.0%] | 1     |
| All ❌✅ (primary)                 | -0.8% | [-3.4%, 1.8%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f5b8a11fecf23cb669f6d1feae81aafc2930aeb9&end=c3f8df39f723e1bae9b21ee801f19e7fc46b1336&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.2%, secondary -0.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 3     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.8%, -0.1%] | 4     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 2     |
| All ❌✅ (primary)                 | -0.2% | [-0.8%, 0.0%]  | 6     |
</details>

**Bootstrap**: 467.825s -> 470.343s (0.54%)
**Artifact size**: 377.36 MiB -> 377.34 MiB (-0.00%)

joshtriplett commented:
> This PR caused some mild perf. regressions across the board. I think we could eat them, provided that this teaches the compiler to do new stuff, but wanted to "let it be known", in case someone has ideas on improving perf. here.

This does teach the compiler some new things, but I have some ideas that might be worth exploring for performance. Top of the list: looking at the generated bitflags code to make sure everything that should be inlined is inlined. Some of the functions are called on hot paths, and they really need to be inlined bit operations, not functions.
